### DISCUSSION TITLE
how to flash ESP32-S2 mini

### DISCUSSION THREAD

**Discussion by gaborfri** [emsesp/EMS-ESP32#1625] [General]
### PROBLEM DESCRIPTION
How does the flashing of an EST32-s2 mini work? 
"No serial data received" appears in the EMS-ESP-Flasher.
if I try it via cmdline, after several attempts I get the message that the bootloader is not S2-compatible.
I was able to get the only successful flash with a tasmota-factory-bin.

do i have to compile it first or is there a special s2 image

gabor

### REQUESTED INFORMATION

_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Searched the problem in [issues](https://github.com/emsesp/EMS-ESP32/issues)
- [x] Searched the problem in [discussions](https://github.com/emsesp/EMS-ESP32/discussions)
- [x ] Searched the problem in the [docs](https://emsesp.github.io/docs/Troubleshooting/)
- [ ] Searched the problem in the [chat](https://discord.gg/3J3GgnzpyT)
- [ ] Provide the output of http://ems-esp.local/api/system :

```lua
  System information output here:


```

### TO REPRODUCE

_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR

_A clear and concise description of what you expected to happen._

### SCREENSHOTS

_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT

_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by MichaelDvP** [comment]
There is no prcompiled bin, you have to comile yourself (there is a pio env for S2-mini). 
But you would not get happy wih S2, it is single core and has less RAM. Spend a few euros and get a S3 with PSRAM.

**Comment by math322** [comment]
For my S2 board i have to press Reset + Button 1 and release Reset to bring the board into flash mode. Also i need to add the 2 lines to the S2-mini env:

board_upload.use_1200bps_touch = false
board_upload.wait_for_upload_port = false

And the S2 (as example ESP32-S2FN4R2) has 2MB PSRAM.

In the command line i use the following parameters:

esptool.py --port /dev/ttyACM0 write_flash 0x1000 bootloader.bin 0x8000 partitions.bin 0x10000 firmware.bin

**Reply by gaborfri** [reply]
it works.
1. edit the platform.ini, add the information from math into the block of [env:lolin_s2_mini]
2. compile the sources running "pio run -e lolin_s2_mini -t upload"

and now I am able to see the new wifi and can connect.
thanks for your help

**Comment by gaborfri** [comment]
now after successful flash, i have the problem: how to connect the pins.
3V and GND is easy, but which pin i need to use for RX and TX on the esp-s2?

in the configuration i found the pins rx=23 and tx=5, but there is no pin 23.
i have only 1-18 and 21 and 33-40.

gabor

**Reply by proddy** [reply]
you can pick any pin, as long as it's connected to the circuit board in the correct way. https://emsesp.github.io/docs/Configuring/#board-profile

**Reply by MichaelDvP** [reply]
The predefined S2-Mini profile uses 
LED: 15
Dallas: 7
TX: 12
RX: 11
Button: 0

**Comment by gaborfri** [comment]
Thank you very much, with the pins I can now see the appropriate outputs in the frontend.
but I don't get any messages in the MQTT explorer.
when I look at the system log I unfortunately see "[mqtt] Publish failed: low memory"

Looks like I need a different ESP32 after all. Which one would you recommend?

**Reply by MichaelDvP** [reply]
You can try to reduce or skip the check here https://github.com/emsesp/EMS-ESP32/blob/1db4a33a1d6ed81c410ab5ac4c816248b0c56cac/src/mqtt.cpp#L595
This is mainly for the esp32 without PSRAM. Never tried what are usefull values for the S2 chip.

If you don't need ethernet the esp32S3 is best, pick a module with more than 4 Mb flash and min 2M PSRAM. The Liligo T7-S3 is a good choice. https://www.lilygo.cc/products/t7-s3

**Reply by MartinP1** [reply]
Seems, that this is adopted meanwhile

```
if (heap_caps_get_free_size(MALLOC_CAP_8BIT) < 60 * 1024) { // checks free Heap+PSRAM
```

**Comment by gaborfri** [comment]
the funy part is, the esp32 s2 mini has 4mb flash an 2mb psram
https://www.wemos.cc/en/latest/s2/s2_mini.html
and the system status shows me the following information:
Heap (Free / Max Alloc) 63 KB / 39 KB
PSRAM (Size / Free) 2.045 KB / 2.034 KB
Flash Chip (Size / Speed) 4.096 KB / 80 MHz
i think i need to edit the 40*1024 and then it should work, lets try

**Reply by MichaelDvP** [reply]
You can set both values very low or remove the check. It is to keep esp32 without PSRAM responsive to web. ESP32 have 512kB SRAM, S2 only 320kB, but the large web chunks are allocated in PSRAM if present. 
In normal operation the limit is not reached (or should not), only if there is a very weak wifi connection and high mqtt publish rate the queue consumes more and more memory and reaches the limit, Sadly the users with weak connection and high rate don't want to change anything on their side and keeps complaining about "freezing" web connection, so we use this way.

**Reply by math322** [reply]
I have set both values to 10 * 1024. Maybe we should move to heap_caps_get_free_size(MALLOC_CAP_8BIT) to get the full free heap size. But i don't know if this function works on Chips without PSRAM.

**Reply by proddy** [reply]
`ESP.getFreeHeap()` is the same as `heap_caps_get_free_size()` so it shouldn't make a difference.

@MichaelDvP  I think all calls to `ESP.getFreeHeap()` and `ESP.getMaxAllocHeap()` should eventually be replaced with `EMSESP::system_.getMaxAllocMem()` and `EMSESP::system_.getHeapMem()` so we're only making these core system calls once in the main loop.

**Reply by math322** [reply]
This is how the output of that ram functions look like (S2 mini):

INFO 175188 61740 41488 36852 - 2094863 2074523 2070155 2064372 - 37888 65536 2094080 2136263

```
LOG_INFO("INFO %d %d %d %d - %d %d %d %d - %d %d %d %d",
             // internal RAM
             ESP.getHeapSize(),
             ESP.getFreeHeap(),
             ESP.getMinFreeHeap(),
             ESP.getMaxAllocHeap(),
             // SPI RAM
             ESP.getPsramSize(),
             ESP.getFreePsram(),
             ESP.getMinFreePsram(),
             ESP.getMaxAllocPsram(),
             // other
             EMSESP::system_.getMaxAllocMem() * 1024,
             EMSESP::system_.getHeapMem() * 1024,
             emsesp::EMSESP::system_.PSram() * 1024,
             heap_caps_get_free_size(MALLOC_CAP_8BIT));
```

**Comment by kami83** [comment]
Hi, i also used math322 description. works fine. But which pins for TX/RX did you used on the Wemos S2 mini board? I found TX=12 and RX=11 is this right? BR kami

**Reply by proddy** [reply]
yes, should be fine

**Comment by MartinP1** [comment]
Hmm, it seems quite complicated....
Had success to build the binary for esp32-s2 mini using devcontainers.

Output binary is  `EMS-ESP-3_8_3-dev_5-ESP32S2-4MB+.bin`.
Is this the only thing needed to flash?

no special `partitions.bin` and `boot-app0.bin`?

If needed, how to build the two, or where to find?

EDIT!

Got it running.... 

After building I found the following in folder `./build/firmware`:

```
-rw-r--r-- 1 martin martin 1698336 Mai 31 13:38 EMS-ESP-3_8_3-dev_5-ESP32S2-4MB+.bin
-rw-r--r-- 1 martin martin      32 Mai 31 13:38 EMS-ESP-3_8_3-dev_5-ESP32S2-4MB+.md5
```
(that was causing my question)

After some searching found this in folder `./.pio/build/s2_4M_P` 

```
martin@martin-D2836-S1:~/github/EMS-ESP32/.pio/build/s2_4M_P$ ls -l *.bin
-rw-r--r-- 1 martin martin   12176 Mai 31 13:36 bootloader.bin
-rw-r--r-- 1 martin martin 1698336 Mai 31 13:38 firmware.bin
-rw-r--r-- 1 martin martin    3072 Mai 31 13:36 partitions.bin
```

this was the esptool invocation that worked for me (adopted from the one @gaborfri posted in this thread on Feb 18, 2024)

```
martin@martin-D2836-S1:~/github/EMS-ESP32/.pio/build/s2_4M_P$ sudo python3 /home/martin/.platformio/packages/tool-esptoolpy/esptool.py --port /dev/ttyACM0 write_flash 0x1000 bootloader.bin 0x8000 partitions.bin 0x10000 firmware.bin

```

The ESP32 s2 mini is already added to my home wifi network...

Blinking frequency of blue LED is reduced

Not connected yet to interface board and my buderus WLW186i

EDIT2:

Up and running now at buderus WLW 186i

Some protocol errors (caused by the weak single core ESP32S2 mini?)

```
2026-06-01T14:11:05.240 W 22: [telegram] Incomplete Rx: CB 00 4B 00 88 00 08 00 90 00 10 00 C4 (crc: C2)
2026-06-01T14:27:28.427 I 23: [mqtt] MQTT connected
2026-06-01T14:33:12.094 W 24: [telegram] Incomplete Rx: 88 00 08 00 A3 (crc: 34)
```

**Comment by MartinP1** [comment]
First experiences:

Flashed Version 3.8.3-dev.5

It seems, that frequently the web server including api stalls. I think that happens when a second client connects to the web server.

MQTT seems to survive the stall.


May be a more performant will resolve the problems. Ordered ESP32S3 with 16 MB flash and 8 MB psram.
